src/portable/st/stm32_hal_driver/hcd_stm32_f7.c

```diff
@@ -0,0 +1,507 @@
+	pipes[i].ep_num = tu_edpt_number(pipes[i].ep_addr);
+	pipes[i].ep_dir = tu_edpt_dir(pipes[i].ep_addr);
+	pipes[i].ep_type = ep_desc->bmAttributes.xfer;
+
+	if (pipes[i].ep_type == EP_TYPE_CTRL && pipes[i].ep_dir == TUSB_DIR_OUT) {
+		/* At a device address, the control endpoint is always number 0, meaning:
+		   control OUT is always pipe number 0, and control IN is always pipe number 1*/
+		pipes[i].num = 0;
+	} else {
+		/* Calculate the pipe number at one device:
+		   taking control OUT pipe numbder as the base */
+		pipes[i].num = i - find_pipe_index_based_on_endpoint_address(pipes[i].dev_addr, 0);
+	}
+
+	if (HAL_HCD_HC_Init(&hhcd, pipes[i].num, pipes[i].ep_addr, pipes[i].dev_addr, USB_OTG_SPEED_HIGH, pipes[i].ep_type,
+	                    ep_desc->wMaxPacketSize) != HAL_OK) {
+		return false;
+	}
+	pipes[i].is_opened = true;
+
+	/* Open a IN pipe for the control endpoint manually since tinyUSB does not do so */
+	if (pipes[i].ep_type == EP_TYPE_CTRL && pipes[i].ep_dir == TUSB_DIR_OUT) {
+		return open_control_in_pipe(pipes[i].dev_addr, pipes[i].ep_num, ep_desc->wMaxPacketSize);
+	}
+
+	return true;
+}
+
+bool hcd_setup_send(uint8_t rhport, uint8_t dev_addr, uint8_t const setup_packet[8])
+{
+	(void) rhport;
+
+	int8_t i = find_pipe_index_based_on_endpoint_address(dev_addr, 0);
+	if (i < 0) {
+		/* Should never happen, if happens it means the pipes are not opened correctly */
+		return false;
+	}
+
+	/* Control OUT pipe is always number 0 on each device */
+	if (HAL_HCD_HC_SubmitRequest(&hhcd, pipes[i].num, TUSB_DIR_OUT, TUSB_XFER_CONTROL, USBH_PID_SETUP,
+	                             (uint8_t *) setup_packet, USBH_SETUP_BUFFER_LENGTH, 0) != HAL_OK) {
+		return false;
+	}
+
+	return true;
+}
+
+bool hcd_edpt_xfer(uint8_t rhport, uint8_t dev_addr, uint8_t ep_addr, uint8_t *buffer, uint16_t buflen)
+{
+	(void) (rhport);
+
+	int8_t i = find_pipe_index_based_on_endpoint_address(dev_addr, ep_addr);
+	if (i < 0) {
+		/* Should never happen, if happens it means we opened interfaces that are not selected */
+		return false;
+	}
+
+	if (pipes[i].ep_dir == TUSB_DIR_IN) {
+		/* Make sure the receiving buffer is cleared before sending a receiving request */
+		memset(buffer, 0, buflen);
+	} else {
+		/* Store the OUT buffer and length in case we need to re-send it in HAL_HCD_HC_NotifyURBChange_Callback */
+		pipes[i].buffer = buffer;
+		pipes[i].buflen = buflen;
+	}
+
+	if (HAL_HCD_HC_SubmitRequest(&hhcd, pipes[i].num, pipes[i].ep_dir, pipes[i].ep_type, USBH_PID_DATA, buffer, buflen,
+	                             0) != HAL_OK) {
+		return false;
+	}
+
+	return true;
+}
+
+bool hcd_edpt_clear_stall(uint8_t rhport, uint8_t dev_addr, uint8_t ep_addr)
+{
+	(void) rhport;
+	(void) dev_addr;
+	(void) ep_addr;
+	return true;
+}
+
+/*--------------------------------------------------------------------+
+ * Weak HAL functions
+ *--------------------------------------------------------------------+*/
+void HAL_HCD_MspInit(HCD_HandleTypeDef *handle)
```

Comment:
This PR looks great, much needed in my opinion, but these two functions (MspInit and Deinit) shouldn't be here, they should be taken care of in board_init() and not part of this driver.